File: OJ/leetcode/33.搜索旋转排序数组.py
#
# @lc app=leetcode.cn id=33 lang=python3
#
# [33] 搜索旋转排序数组
#
# https://leetcode-cn.com/problems/search-in-rotated-sorted-array/description/
#
# algorithms
# Medium (36.45%)
# Likes:    561
# Dislikes: 0
# Total Accepted:    84.4K
# Total Submissions: 231.2K
# Testcase Example:  '[4,5,6,7,0,1,2]\n0'
#
# 假设按照升序排序的数组在预先未知的某个点上进行了旋转。
# 
# ( 例如，数组 [0,1,2,4,5,6,7] 可能变为 [4,5,6,7,0,1,2] )。
# 
# 搜索一个给定的目标值，如果数组中存在这个目标值，则返回它的索引，否则返回 -1 。
# 
# 你可以假设数组中不存在重复的元素。
# 
# 你的算法时间复杂度必须是 O(log n) 级别。
# 
# 示例 1:
# 
# 输入: nums = [4,5,6,7,0,1,2], target = 0
# 输出: 4
# 
# 
# 示例 2:
# 
# 输入: nums = [4,5,6,7,0,1,2], target = 3
# 输出: -1
# 
#

# @lc code=start
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        if not nums:
            return -1
        left=0;right=len(nums)-1
        while left<=right:      #注意等号.eg: length=1时
            mid=(left+right)//2
            if target==nums[mid]:
                return mid
            elif nums[left]<=nums[mid]: #在left到mid区间上连续
                if nums[left]<=target and target<=nums[mid]:
                    right=mid-1 #收缩后的区间不包括mid
                else:   
                    left=mid+1  #收缩后的区间不包括mid
            else:                       #在mid到right区间上连续
                if nums[mid]<=target and target<=nums[right]:
                    left=mid+1  #收缩后的区间不包括mid
                else:
                    right=mid-1 #收缩后的区间不包括mid
        
        return -1
# @lc code=end

File: OJ/leetcode/33.搜索旋转排序数组.cpp
// @before-stub-for-debug-begin
#include <vector>
#include <string>
#include "commoncppproblem33.h"

using namespace std;
// @before-stub-for-debug-end

/*
 * @lc app=leetcode.cn id=33 lang=cpp
 *
 * [33] 搜索旋转排序数组
 */

 // @lc code=start
class Solution {
public:

    int search(vector<int>& nums, int target) {
        int n = nums.size();
        int l = 0, r = n - 1;
        while (l <= r)
        {
            int mid = l + (r - l) / 2;
            if (nums[mid] == target)
                return mid;
            //去除两边，逐渐缩短
            if (nums[0] <= nums[mid]) {
                //当target位于nums[0]和nums[mid]之间时
                if (nums[0] <= target && target < nums[mid]) {
                    //往左找
                    r = mid - 1;
                }
                //其余往右找，去除左端
                else {
                    l = mid + 1;
                }
            }
            else {
                //当target位于nums[mid]和nums[n-1]之间时
                if (nums[mid] < target && target <= nums[n - 1]) {
                    //往右找
                    l = mid + 1;
                }
                //其余往左找，去除右端
                else {
                    r = mid - 1;
                }
            }
        }
        return -1;


    }
};
// @lc code=end

